src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts

```diff
@@ -156,6 +184,140 @@ export class CommandDetectionCapability implements ICommandDetectionCapability {
+			console.log('ITerminalCommand', e);
+
+			// Search for start of term
+			if (!e.marker?.line) {
+				return;
+			}
+			const line = this._terminal.buffer.active.getLine(e.marker.line);
+			if (!line) {
+				return;
+			}
+			const term = 'gti';
+			let start = 0;
+			let letter = 0;
+			let cell: IBufferCell | undefined;
+			for (let i = 0; i < line.length; i++) {
+				cell = line.getCell(i, cell);
+				if (!cell) {
+					continue;
+				}
+				if (cell.getChars() === term[letter]) {
+					if (letter === 0) {
+						start = i;
+					}
+					letter++;
+					if (letter === term.length) {
+						break;
+					}
+				} else {
+					letter = 0;
+				}
+				// if (line.getCell(start)?.getChars();
+			}
+			if (letter === term.length) {
+				// TODO: pass in action which writes to the terminal the correct term
+				console.log(`Found "${term}"! line: ${e.marker.line}, start: ${start}`);
+				this._registerContextualDecoration(e.marker, undefined, start, term.length, true);
+			}
+			// console.log('output', e.getOutput());
+		});
+
+		this.registerCommandFinishedListener(/.+/, command => {
+			const output = command.getOutput();
+			//TODO: support this for tasks
+			//TODO: narrow this down more
+			// if (command.exitCode === 1) {
+			// TODO: add other possible regexps
+			const match = output?.match(/.*address already in use \d\.\d.\d\.\d:(\d\d\d\d).*/);
+			if (match && match.length === 2 && command.endMarker) {
+				console.log(`address in use for port ${match[1]}`);
+				const actions: IAction[] = [];
+				const labelRun = localize("terminal.freePort", 'Free port');
+				actions.push({
+					class: undefined, tooltip: labelRun, id: 'terminal.freePort', label: labelRun, enabled: true,
+					run: async () => {
+						const procs = await new Promise<string>((resolve, reject) => {
+							cp.exec(`lsof -nP -iTCP -sTCP:LISTEN | grep ${match[1]}`, {}, (err, stdout) => {
+								if (err) {
+									return reject('Problem occurred when listing active processes');
+								}
+								resolve(stdout);
+							});
+						});
+						const procLines = procs.split('\n');
+						if (procLines.length >= 1) {
+							const regex = /\s+(\d+)\s+/;
+							const matches = procLines[0].match(regex);
+							if (matches && matches.length > 1) {
+								console.log('found pid', matches[1]);
+								await new Promise<string>((resolve, reject) => {
+									cp.exec(`kill ${matches[1]}`, {}, (err, stdout) => {
+										if (err) {
+											console.log('error', err);
+											return reject(`Problem occurred when killing the process w id: ${matches[1]}`);
+										}
+										console.log(`killed process ${matches[1]}`);
+										resolve(stdout);
+									});
+								});
+							}
+						}
+					}
+				});
+
+				this._registerContextualDecoration(command.endMarker, actions);
+			}
+			// }
+		});
+	}
+
+	private _registerContextualDecoration(marker: IMarker, actions?: IAction[], x?: number, width?: number, border?: boolean): void {
```

Comment:
The decoration stuff would be better to live outside of the capability (this might have been like this in my rough exploration)